### PR TITLE
Implement reference counting

### PR DESCRIPTION
**Pull request by dcodeIO**
As a continuation of the runtime branch. this is the new tracking PR for everything runtime and RC. What this essentially does is implement and integrate reference counting, best explained by [this test case](https://github.com/AssemblyScript/assemblyscript/blob/dev/tests/compiler/retain-release.ts) resulting in [this output](https://github.com/AssemblyScript/assemblyscript/blob/dev/tests/compiler/retain-release.untouched.wat).

Changes to master are significant in that

* all runtime information is now stored in a single runtime header shared by both the allocator and the reference counting implementation.
  ```ts
  // ╒════════════════ Common block layout (32-bit) ═════════════════╕
  //    3                   2                   1
  //  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0  bits
  // ├─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┤
  // │                           MM info                             │ -16
  // ├───────────────────────────────────────────────────────────────┤
  // │                           GC info                             │ -12
  // ├───────────────────────────────────────────────────────────────┤
  // │                          runtime id                           │ -8
  // ├───────────────────────────────────────────────────────────────┤
  // │                         runtime size                          │ -4
  // ╞═══════════════════════════════════════════════════════════════╡
  // │                              ...                              │ ref
  ```
* ArrayBuffer and String don't expose the runtime header (or even .byteLength, .length) anymore but their data starts right at their reference, which is achieved by relocating the header as mentioned above. In turn, previous helper macros like `LOAD` are gone because one can now simply do a `load` instead.
* a user cannot pick a memory allocator anymore by means of `import "..."` but instead has to select an appropriate runtime implementation. By default that's `full`, which is essentially TLSF + PureRC, `half` which is the same but without exported runtime hooks, `stub` which is essentially Arena + no GC and `none` which is the same without exported runtime hooks. The none implementation, even though runtime hooks are still inserted by the compiler, essentially evaporates after optimizations. Also, though unlikely, custom implementations are possible by means of `--runtime yourRuntimeImplementation` which implements all the runtime hooks.
* the memory allocator and collector interfaces are gone and there's no `memory.allocate` and `memory.free` anymore. When done, the recommended way of allocating a chunk of raw memory will be `new ArrayBuffer`.
* the previous interchangeable GC integration is gone because it turned out that trying to support both tracing and reference counting interchangeably is insane.
* dynamic `instanceof` (i.e. `animal instanceof Cat`) is implemented on top of runtime type information, that is embedded as a memory segment. For each class there are 4 bytes base class id + 4 bytes flags with alignments, managed status etc..
* guaranteed alignment is now 16 bytes to properly fit up to v128 values.
* reallocation is done on memory manager level.
* various TLSF issues have been fixed and there's now a visual debugger with a variety of buttons to press.
* nullability checks have been implemented in flows, so the compiler can now infer where a reference is guaranteed to be non-null or emit an error otherwise.
* standard library is not optional anymore, so `--noLib` is gone. This significantly reduces the amount of checks the compiler has to do when doing anything standard library related. This doesn't matter effectively, because if one doesn't use let's say Strings nothing of the related library components becomes compiled anyway.
* compiler-level tree shaking is always on now (equals previous default behavior), so `--noTreeShaking` is gone. We'll eventually want to replace this with a pre-pass checker anyway.
* the function table is not exported anymore in order to allow Binaryen to do all its optimizations, including converting indirect calls to direct calls after other passes.

There's probably quite a lot more that I don't remember anymore because this has clearly taken too long already. The test case linked above gives an overview of what's still to do on the integration side, plus I have yet to test the actual "full" runtime implementation. The concept should be figured out now, though.

*to be continued...*

### CODE REVIEW

**Comment by jtenner** [comment]
Aspect needs wasm. getFunction() from the loader in order to work. Will this be supported behind a compiler flag?

**Comment by dcodeIO** [comment]
Can make this an `--exportTable` option, yeah.

**Comment by jtenner** [comment]
I'm forgetting that I can export a function that uses `call_indirect` directly.

**Comment by dcodeIO** [comment]
Alright, committed my progress so far. Some things seem to be working, like std/array and std/string, others seem to be completely broken, like std/array-access that emits rogue releases everywhere. Good enough for a first glimpse I hope :)

**Comment by MaxGraey** [comment]
Happy to see this great progress! Well done!

But also I see some regression from updated binaryen like here:

was: https://github.com/AssemblyScript/assemblyscript/pull/592/files/f73d807d5a893989608613a9f0a367ff117fb980..3e480d9423658b805c10fab2a8467cfa6e241191#diff-587a89b754ca1898d922a9e2c289bcfcL131

now: https://github.com/AssemblyScript/assemblyscript/pull/592/files/f73d807d5a893989608613a9f0a367ff117fb980..3e480d9423658b805c10fab2a8467cfa6e241191#diff-587a89b754ca1898d922a9e2c289bcfcR160

It seems someting with constant propagation pass

**Comment by dcodeIO** [comment]
I've disabled the optimizer when creating the tests so it doesn't optimize away my precious retains. Could be the cause. Thinking of updating the test suite to run both the unoptimized and optimized versions so we can have both files again.

**Comment by MaxGraey** [comment]
Oh, that's explain "regression"

**Comment by dcodeIO** [comment]
Last commit should have proper optimized fixtures again, unless I missed one. Test suite now runs both.

**Comment by dcodeIO** [comment]
Some thoughts:
* The loader doesn't support dealing with the new arrays yet and there are multiple ways this could be done now. One is to give all the basic views static ids, similar to ArrayBuffer (0) and String (1), but that would mean that the `__visit_members` builtin grows a bit even if these aren't used because these are always in RTTI. The other is to move all of that to the users, requiring them to export `idof<T>()` of whatever they want to use.
* While the tests pass now, and the RTrace thing properly validates that there's an even number of retain and release calls on each object (except heap objects which are ignored by RC), it's hard to tell whether the underlying implementation actually does what it is supposed to. Relatively confident though that it doesn't corrupt memory.
* Optimizing the RC calls is harder now than it should be because I had to use some not-so-great helpers like `__retainRelease` and `__skippedRelease`. My hope initially was that there could be a Binaryen pass based on `__retain` and `__release` calls alone, condensing these, but with all the special stuff this seems out of question. Currently investigating alternatives, like traversing whatever expression is retained/released backwards to blacklist local ids we can't use for a temp. ([see also](https://github.com/WebAssembly/binaryen/issues/2135)).

Other than that this _should_ be good for test drives now. 🚀

**Comment by jtenner** [comment]
@dcodeIO if this is good for test driving, could it be possible to get the dist files updated here? Or should I just fork this branch and update dist myself?

**Comment by dcodeIO** [comment]
Need to configure CI to exclude this branch/PR from no-changes-to-dist checks somehow. Going to take a look :)

**Comment by dcodeIO** [comment]
Interestingly, `-O3s` (optimizeLevel=3, shrinkLevel=1) sometimes yields considerably larger outputs than just `-O3` (optimizeLevel3, shrinkLevel=0). [+261](https://github.com/AssemblyScript/assemblyscript/pull/592/commits/af00bdeefe9090f34ce3a98df1060d02ee3377ec#diff-57e1922c476174aeb3be8faf0b466c87), [+11/start](https://github.com/AssemblyScript/assemblyscript/pull/592/commits/af00bdeefe9090f34ce3a98df1060d02ee3377ec#diff-c7ccd981821b24ff480c2f5c1e1f717c), [+174](https://github.com/AssemblyScript/assemblyscript/pull/592/commits/af00bdeefe9090f34ce3a98df1060d02ee3377ec#diff-68597130a81695edba1a7e2c4cce1d92), [+76](https://github.com/AssemblyScript/assemblyscript/pull/592/commits/af00bdeefe9090f34ce3a98df1060d02ee3377ec#diff-c4954a4a4e58d1adaab510feaeac6f87), [+1040](https://github.com/AssemblyScript/assemblyscript/pull/592/commits/af00bdeefe9090f34ce3a98df1060d02ee3377ec#diff-0717d038bca8614e303eb6c10b9ccafa) @kripken maybe?

**Comment by kripken** [comment]
That's somewhat surprising - do you have a wast file that when I run it with -O3 versus -O3 + shrink 1 then the second is larger?

**Comment by dcodeIO** [comment]
Maybe this one: [builtins.untouched.wat](https://github.com/AssemblyScript/assemblyscript/blob/af00bdeefe9090f34ce3a98df1060d02ee3377ec/tests/compiler/builtins.untouched.wat). Seems that shrinkLevel=1 skips some inlining steps, no matter the optimizeLevel.

**Comment by dcodeIO** [comment]
For anyone interested, there's [a new loader API](https://github.com/AssemblyScript/assemblyscript/tree/dev/lib/loader).

**Comment by kripken** [comment]
I see, thanks @dcodeIO - yeah, when shrink mode is on, we are more careful with inlining. So we don't inline functions we are not absolutely sure will end up in smaller code size. In this example, inlining them does help - e.g. the first seems to be an isnan, and called with an argument that is a constant, so it ends up constant anyhow. But if the argument were not a constant, code size would increase. So we'd need to add more sophisticated decision-making here to handle that.  I opened https://github.com/WebAssembly/binaryen/issues/2144

**Comment by dcodeIO** [comment]
So, the solution to the stack limitations I utilize now is an expression walker on our end that blacklists local indexes used in another expression, but this can potentially also be used for other stuff now that we have it,, like AssemblyScript-specific optimizations. Will see how this turns out :)

**Comment by dcodeIO** [comment]
With WASM Studio currently broken, I think we should consider merging dev->master rather sooner than later and continue from there. It's very likely that there will be issues, but as always these will surface the sooner the more people try 0.7. With current master having its own issues this won't make much of a difference I guess. Wdyt?

**Comment by MaxGraey** [comment]
Personally I fully agree. This will make it possible to test new features much faster. It would be so nice to start publishing on npm and follow semver and bump minors versions as well

**Comment by MaxGraey** [comment]
About last PR. It seems I forget rebuild tests for simd. @dcodeIO could you force this?

**Comment by dcodeIO** [comment]
Sure, will do with my next commit :)

File: std/assembly/array.ts [hunk 1]
```diff
@@ -1,148 +1,156 @@
...
-          <u8>value,
+          dataStart + <usize>start,
+          u8(value),
           <usize>(end - start)
         );
       }
     } else {
       for (; start < end; ++start) {
-        STORE<T>(buffer, start, value);
+        store<T>(dataStart + (<usize>start << alignof<T>()), value);
```
Comment (MaxGraey) [inline std/assembly/array.ts]: I guess this need check `isManged<T>()` and insert `__gc_link`. wdyt?

Comment (dcodeIO) [reply std/assembly/array.ts]: Good find, same seems to be the case for copyWithin

**Pull request merged**